File: packages/sdk/python/mcp-server/src/mcp_obs_server/oauth_middleware.py
"""
OAuth middleware utilities for MCP servers
Provides decorator functions to add OAuth authentication to MCP request handlers
"""
import time
from typing import Any, Callable, Optional, List, Dict, Union, Awaitable
from functools import wraps
from pydantic import BaseModel
from loguru import logger

from .types import OAuthConfig, AuthContext
from .oauth_validator import OAuthTokenValidator


class MCPError(Exception):
    """MCP protocol error with error code"""

    def __init__(self, code: int, message: str):
        super().__init__(message)
        self.code = code


class OAuthMiddlewareConfig(OAuthConfig):
    """Extended OAuth configuration for middleware"""
    required_scopes: List[str] = []
    """Required scopes for all protected endpoints"""

    skip_validation_for: List[str] = []
    """Skip OAuth validation for specific tools"""


# Type definitions
AuthenticatedHandler = Callable[[Any, AuthContext], Awaitable[Any]]
RequestHandler = Callable[[Any], Awaitable[Any]]

# Global validator instance
_global_validator: Optional[OAuthTokenValidator] = None


def configure_oauth_validator(config: OAuthMiddlewareConfig) -> None:
    """Configure global OAuth validator instance"""
    global _global_validator
    _global_validator = OAuthTokenValidator(config)


def get_oauth_validator() -> OAuthTokenValidator:
    """Get the configured OAuth validator instance"""
    if _global_validator is None:
        raise RuntimeError("OAuth validator not configured. Call configure_oauth_validator() first.")
    return _global_validator


def with_oauth(
    handler: AuthenticatedHandler,
    *,
    required_scopes: Optional[List[str]] = None,
    skip_validation_for: Optional[List[str]] = None
) -> RequestHandler:
    """
    Decorator that wraps MCP request handlers with OAuth authentication

    Args:
        handler: The original request handler that expects auth context
        required_scopes: Optional list of required scopes for this handler
        skip_validation_for: Optional list of tool names to skip validation for

    Returns:
        Wrapped handler that validates OAuth before calling original handler
    """
    @wraps(handler)
    async def wrapper(request: Any) -> Any:
        validator = get_oauth_validator()

        # Check if we should skip validation for this specific tool
        tool_name = getattr(getattr(request, 'params', None), 'name', None)
        if skip_validation_for and tool_name in skip_validation_for:
            # Call handler without auth context for skipped tools
            anonymous_context = AuthContext(
                user_id="anonymous",
                email="anonymous@mcp-obs.com",
                scopes=[],
                client_id="anonymous",
                expires_at=int(time.time() * 1000) + 3600000  # 1 hour from now
            )
            return await handler(request, anonymous_context)

        # Extract Bearer token from request
        token = _extract_token_from_request(request)

        if not token:
            raise MCPError(
                code=-32602,  # Invalid params error code
                message="Authorization required. Include Bearer token in Authorization header."
            )

        # Validate token
        auth_context = await validator.validate_token(token)

        if not auth_context:
            raise MCPError(
                code=-32602,  # Invalid params error code
                message="Invalid or expired OAuth token"
            )

        # Check required scopes
        if required_scopes and not validator.has_required_scopes(auth_context, required_scopes):
            raise MCPError(
                code=-32603,  # Internal error code (closest to 403 Forbidden)
                message=f"Insufficient scopes. Required: {', '.join(required_scopes)}, "
                       f"Got: {', '.join(auth_context.scopes)}"
            )

        # Call original handler with auth context
        return await handler(request, auth_context)

    return wrapper


def with_oauth_stdio(config: OAuthMiddlewareConfig) -> Callable[[AuthenticatedHandler], RequestHandler]:
    """
    Wrapper for stdio transport OAuth middleware
    Handles Bearer token extraction from MCP request metadata

    Args:
        config: OAuth middleware configuration

    Returns:
        Decorator function for stdio transport
    """
    # Configure validator if not already done
    if _global_validator is None:
        configure_oauth_validator(config)

    def decorator(handler: AuthenticatedHandler) -> RequestHandler:
        return with_oauth(
            handler,
            required_scopes=config.required_scopes,
            skip_validation_for=config.skip_validation_for
        )

    return decorator


def with_oauth_http(config: OAuthMiddlewareConfig) -> Callable[[AuthenticatedHandler], RequestHandler]:
    """
    Wrapper for HTTP transport OAuth middleware
    Handles Bearer token extraction from HTTP Authorization headers

    Args:
        config: OAuth middleware configuration

    Returns:
        Decorator function for HTTP transport
    """
    # Configure validator if not already done
    if _global_validator is None:
        configure_oauth_validator(config)

    def decorator(handler: AuthenticatedHandler) -> RequestHandler:
        return with_oauth(
            handler,
            required_scopes=config.required_scopes,
            skip_validation_for=config.skip_validation_for
        )

    return decorator


def _extract_token_from_request(request: Any) -> Optional[str]:
    """
    Extract Bearer token from MCP request
    This is a transport-agnostic extraction that tries multiple methods

    Args:
        request: MCP request object

    Returns:
        Bearer token string or None if not found
    """
    validator = get_oauth_validator()

    # Method 1: Check if request has an Authorization header (HTTP transports)
    headers = getattr(request, 'headers', None)
    if headers:
        # Try both lowercase and uppercase
        auth_header = headers.get('authorization') or headers.get('Authorization')
        if auth_header:
            token = validator.extract_bearer_token(auth_header)
            if token:
                return token

    # Method 2: Check MCP metadata (stdio transport)
    metadata = getattr(request, 'metadata', None)
    if metadata:
        auth_header = metadata.get('authorization') or metadata.get('Authorization')
        if auth_header:
            token = validator.extract_bearer_token(auth_header)
            if token:
                return token

    # Method 3: Check request params for authorization (fallback)
    params = getattr(request, 'params', None)
    if params:
        auth_header = getattr(params, 'authorization', None)
        if auth_header:
            token = validator.extract_bearer_token(auth_header)
            if token:
                return token

    # Method 4: Check top-level authorization field (custom transport)
    auth_header = getattr(request, 'authorization', None)
    if auth_header:
        token = validator.extract_bearer_token(auth_header)
        if token:
            return token

    return None


def create_oauth_challenge_headers(
    server_slug: str,
    server_id: Optional[str] = None,
) -> Dict[str, str]:
    """
    Utility function to create OAuth challenge response headers
    For HTTP transports that need WWW-Authenticate header

    Args:
        server_slug: MCP server slug
        server_id: Optional server ID for advanced header generation

    Returns:
        Dictionary of response headers
    """
    # Fallback to simple challenge header
    protocol = "http" if "localhost" in server_slug else "https"
    base_url = server_slug if "localhost" in server_slug else f"{server_slug}.mcp-obs.com"

    return {
        "WWW-Authenticate": f'Bearer resource_metadata="{protocol}://{base_url}/.well-known/oauth-protected-resource"'
    }


def is_authenticated(request: Any) -> bool:
    """
    Utility to check if request is authenticated
    Useful for conditional logic in handlers

    Args:
        request: MCP request object

    Returns:
        True if request has a valid Bearer token
    """
    try:
        token = _extract_token_from_request(request)
        return token is not None
    except Exception:
        return False


def get_request_correlation_id(request: Any) -> str:
    """
    Utility to get correlation ID for request tracking
    Generates a unique ID for each request for logging/debugging

    Args:
        request: MCP request object

    Returns:
        Correlation ID string
    """
    # Try to get existing correlation ID
    existing = getattr(request, 'correlation_id', None) or getattr(request, 'id', None)
    if existing:
        return str(existing)

    # Generate new correlation ID
    timestamp = int(time.time() * 1000)
    random_suffix = hex(hash(str(timestamp) + str(id(request))))[-9:]
    return f"req_{timestamp}_{random_suffix}"

File: packages/sdk/typescript/mcp-server/src/oauth-middleware.ts
/**
 * OAuth middleware utilities for MCP servers
 * Provides wrapper functions to add OAuth authentication to MCP request handlers
 */

import type { Request as MCPRequest, ErrorCode } from "@modelcontextprotocol/sdk/types.js";
import { OAuthTokenValidator, type AuthContext, type OAuthConfig } from "./oauth-validator.js";

export interface MCPError extends Error {
  code: ErrorCode;
}

// Create MCP Error constructor
function createMCPError(code: ErrorCode, message: string): MCPError {
  const error = new Error(message) as MCPError;
  error.code = code;
  return error;
}

export interface OAuthMiddlewareConfig extends OAuthConfig {
  /** Required scopes for all protected endpoints */
  requiredScopes?: string[];
  /** Skip OAuth validation for specific tools */
  skipValidationFor?: string[];
  /** Support tool configuration */
  supportTool?: {
    enabled?: boolean;
    title?: string;
    description?: string;
    categories?: string[];
  };
}

export type AuthenticatedHandler<T extends MCPRequest> = (
  request: T,
  authContext: AuthContext
) => Promise<any>;

export type RequestHandler<T extends MCPRequest> = (request: T) => Promise<any>;

let globalValidator: OAuthTokenValidator | null = null;

/**
 * Configure global OAuth validator instance
 */
export function configureOAuthValidator(config: OAuthMiddlewareConfig): void {
  globalValidator = new OAuthTokenValidator(config);
}

/**
 * Get the configured OAuth validator instance
 */
export function getOAuthValidator(): OAuthTokenValidator {
  if (!globalValidator) {
    throw new Error('OAuth validator not configured. Call configureOAuthValidator() first.');
  }
  return globalValidator;
}

/**
 * Higher-order function that wraps MCP request handlers with OAuth authentication
 * @param handler The original request handler that expects auth context
 * @param options Optional configuration for this specific handler
 */
export function withOAuth<T extends MCPRequest>(
  handler: AuthenticatedHandler<T>,
  options?: {
    requiredScopes?: string[];
    skipValidationFor?: string[];
  }
): RequestHandler<T> {
  return async (request: T): Promise<any> => {
    const validator = getOAuthValidator();

    // Check if we should skip validation for this specific tool
    if (options?.skipValidationFor?.includes((request.params as any)?.name)) {
      // Call handler without auth context for skipped tools
      return handler(request, {
        userId: 'anonymous',
        email: 'anonymous@mcp-obs.com',
        scopes: [],
        clientId: 'anonymous',
        expiresAt: Date.now() + 3600000 // 1 hour from now
      });
    }

    // Extract Bearer token from request
    const token = extractTokenFromRequest(request);

    if (!token) {
      throw createMCPError(
        -32602, // Invalid params error code
        'Authorization required. Include Bearer token in Authorization header.'
      );
    }

    // Validate token
    const authContext = await validator.validateToken(token);

    if (!authContext) {
      throw createMCPError(
        -32602, // Invalid params error code
        'Invalid or expired OAuth token'
      );
    }

    // Check required scopes
    const requiredScopes = options?.requiredScopes || [];
    if (requiredScopes.length > 0 && !validator.hasRequiredScopes(authContext, requiredScopes)) {
      throw createMCPError(
        -32603, // Internal error code (closest to 403 Forbidden)
        `Insufficient scopes. Required: ${requiredScopes.join(', ')}, Got: ${authContext.scopes.join(', ')}`
      );
    }

    // Call original handler with auth context
    return handler(request, authContext);
  };
}

/**
 * Wrapper for stdio transport OAuth middleware
 * Handles Bearer token extraction from MCP request metadata
 */
export function withOAuthStdio<T extends MCPRequest>(
  config: OAuthMiddlewareConfig
): (handler: AuthenticatedHandler<T>) => RequestHandler<T> {
  // Configure validator if not already done
  if (!globalValidator) {
    configureOAuthValidator(config);
  }

  return (handler: AuthenticatedHandler<T>) => {
    return withOAuth(handler, {
      requiredScopes: config.requiredScopes,
      skipValidationFor: config.skipValidationFor
    });
  };
}

/**
 * Wrapper for HTTP transport OAuth middleware
 * Handles Bearer token extraction from HTTP Authorization headers
 */
export function withOAuthHTTP<T extends MCPRequest>(
  config: OAuthMiddlewareConfig
): (handler: AuthenticatedHandler<T>) => RequestHandler<T> {
  // Configure validator if not already done
  if (!globalValidator) {
    configureOAuthValidator(config);
  }

  return (handler: AuthenticatedHandler<T>) => {
    return withOAuth(handler, {
      requiredScopes: config.requiredScopes,
      skipValidationFor: config.skipValidationFor
    });
  };
}

/**
 * Extract Bearer token from MCP request
 * This is a transport-agnostic extraction that tries multiple methods
 */
function extractTokenFromRequest(request: MCPRequest): string | null {
  // Try to extract from different possible locations

  // Method 1: Check if request has an Authorization header (HTTP transports)
  const headers = (request as any).headers;
  if (headers && headers.authorization) {
    return getOAuthValidator().extractBearerToken(headers.authorization);
  }
  if (headers && headers.Authorization) {
    return getOAuthValidator().extractBearerToken(headers.Authorization);
  }

  // Method 2: Check MCP metadata (stdio transport)
  const metadata = (request as any).metadata;
  if (metadata && metadata.authorization) {
    return getOAuthValidator().extractBearerToken(metadata.authorization);
  }
  if (metadata && metadata.Authorization) {
    return getOAuthValidator().extractBearerToken(metadata.Authorization);
  }

  // Method 3: Check request params for authorization (fallback)
  const params = (request as any).params;
  if (params && params.authorization) {
    return getOAuthValidator().extractBearerToken(params.authorization);
  }

  // Method 4: Check top-level authorization field (custom transport)
  if ((request as any).authorization) {
    return getOAuthValidator().extractBearerToken((request as any).authorization);
  }

  return null;
}

/**
 * Utility function to create OAuth challenge response headers using injected function
 * For HTTP transports that need WWW-Authenticate header
 */
export async function createOAuthChallengeHeaders(
  serverSlug: string,
  serverId?: string,
  generateWWWAuthenticateHeaderFn?: (serverId: string, realm?: string, error?: string, errorDescription?: string) => Promise<string>
): Promise<Record<string, string>> {
  try {
    if (serverId && generateWWWAuthenticateHeaderFn) {
      // Use injected WWW-Authenticate header generation function
      const wwwAuth = await generateWWWAuthenticateHeaderFn(
        serverId,
        'MCP Server',
        'invalid_token',
        'Bearer token required'
      );
      return {
        'WWW-Authenticate': wwwAuth
      };
    }
  } catch (error) {
    console.error('[OAuth] Error generating WWW-Authenticate header:', error);
  }

  // Fallback to simple challenge header
  const protocol = serverSlug.includes('localhost') ? 'http' : 'https';
  const baseUrl = serverSlug.includes('localhost') ? serverSlug : `${serverSlug}.mcp-obs.com`;

  return {
    'WWW-Authenticate': `Bearer resource_metadata="${protocol}://${baseUrl}/.well-known/oauth-protected-resource"`
  };
}

/**
 * Utility to check if request is authenticated
 * Useful for conditional logic in handlers
 */
export function isAuthenticated(request: MCPRequest): boolean {
  try {
    const token = extractTokenFromRequest(request);
    return token !== null;
  } catch {
    return false;
  }
}

/**
 * Utility to get correlation ID for request tracking
 * Generates a unique ID for each request for logging/debugging
 */
export function getRequestCorrelationId(request: MCPRequest): string {
  // Try to get existing correlation ID
  const existing = (request as any).correlationId || (request as any).id;
  if (existing) {
    return existing.toString();
  }

  // Generate new correlation ID
  return `req_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
}

/**
 * Configure MCP server with OAuth middleware and optional support tool
 * This is the main integration function that sets up both authentication and support
 */
export async function configureOAuthMCPServer(
  server: any, // MCP Server instance
  config: OAuthMiddlewareConfig
): Promise<void> {
  // Configure OAuth validator
  configureOAuthValidator(config);

  // Register support tool if enabled
  if (config.supportTool?.enabled) {
    const { registerSupportTool } = await import('./support-tool.js');

    await registerSupportTool(server, {
      enabled: true,
      title: config.supportTool.title,
      description: config.supportTool.description,
      categories: config.supportTool.categories,
      serverSlug: config.serverSlug,
      serverId: config.serverId
    });

    // Add support tool to skip validation list if not already there
    if (!config.skipValidationFor?.includes('get_support_tool')) {
      config.skipValidationFor = [...(config.skipValidationFor || []), 'get_support_tool'];
    }
  }

  console.log('✅ MCP Server configured with OAuth middleware' +
    (config.supportTool?.enabled ? ' and support tool' : ''));
}